### PR TITLE
Fix: Creating new Loop pages after SharePoint domain change

### PR DESCRIPTION
**Pull request by Philip-Scott**
## Description
After a tenant changes their domain, this lead existing workspaces on the Loop app to stop allowing the creation of new pages. After a redirect, the authorization headers are not sent by the browser. And even if were to just retry with the headers set, SharePoint requires for the auth tokens to be generated for that particular domain. 

After these changes, I was able to get new pages to attach again. What I did was catch the redirect on the `fetchHelper(` function, and allow getFileLink to handle the redirect for us. For this to work, we need to request new tokens for the new domain.  

image

The fix is sadly not perfect as we need to go through the following flow in order to detect and retry the redirect error. 

- The first call to getShareURL against the original URL gives us the 308
- The second one that gets made automatically by the browser to follow the 308 will be made against the new domain, but will fail due to tokens
- We need to catch that second one, and retry with new tokens. This call will succeed

### CODE REVIEW

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 1]
```diff
@@ -106,42 +153,34 @@ async function getFileLinkCore(
...
-						"Accept": "application/json;odata=verbose",
-						...headers,
-					},
-				};
-				const response = await fetchHelper(url, requestInit);
+				let response = await fetchHelper(url, requestInit);
+
+				// Handle redirects by requesting new tokens & retrying the API call against the new site URL.
+				if (!response.content.ok && response.content.redirected) {
+					odspUrlParts.siteUrl = response.content.url.split("/_api/")[0];
```
Comment (Philip-Scott) [inline packages/drivers/odsp-driver/src/getFileLink.ts:169]: Do we have a known way to get the siteURL from a URL? That would avoid the need to requiring me to split the URL as I did here

File: packages/drivers/odsp-driver/src/odspUtils.ts [hunk 2]
```diff
@@ -122,6 +122,18 @@ export async function fetchHelper(
 					{ driverVersion },
 				);
 			}
+
+			if (!response.ok && response.redirected) {
+				// Return redirected response to allow callers of fetchHelper to handle the redirect.
+				const redirectHeaders = headersToMap(response.headers);
+				return {
+					content: response,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/odspUtils.ts:130]: This does not look right. If it's an error, we should throw here. Other consumers (other than getFileLinkCore) will expect success path to be in a format they expect, not an error response here.
I think we should craft exception with a proper information that is sufficient for redirect flow, but not return success.

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 3]
```diff
@@ -106,42 +153,34 @@ async function getFileLinkCore(
...
+
+					const redirectedResponseInformation = await getRequestInformation(
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/getFileLink.ts:171]: @jatgarg is a much better person to comment on that flow, but I'd expect that any pre-step and after-step to be reorganized into helper function that could be used on any other code path that require to be enlighten about redirect flow (i.e. code reuse). Maybe even in the form of retryWithRedirect(callback) kind of helper functionality.

File: packages/loader/driver-utils/src/index.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ export {
 	getRetryDelaySecondsFromError,
 	isOnline,
 	LocationRedirectionError,
+	RequestRedirectionError,
```
Comment (jatgarg) [inline packages/loader/driver-utils/src/index.ts:25]: Why this new error type. We already have the error type for this. "LocationRedirectionError". We should use that only and not create duplicate error types.

File: packages/drivers/odsp-driver/src/odspUtils.ts [hunk 5]
```diff
@@ -122,6 +123,13 @@ export async function fetchHelper(
 					{ driverVersion },
 				);
 			}
+
+			if (!response.ok && response.redirected) {
+				throw new RequestRedirectionError("ODSP fetch error: Redirection", response.url, {
```
Comment (jatgarg) [inline packages/drivers/odsp-driver/src/odspUtils.ts:128]: I think this could interrupt our other flows. This is just a fetch util and we don't create these types of errors here. They are generated in "throwOdspNetworkError" utility.

File: packages/loader/driver-utils/src/index.ts (hunk 4 above)
Comment (Philip-Scott) [reply packages/loader/driver-utils/src/index.ts:25]: I had to make a new error type for this redirect as the previously existing LocationRedirectionError was asking for things from IResolvedUrl which I did not have access to. I could just "fake" the data from IResolvedUrl that I don't have access to if you think that  is better than having a new error

Comment (Philip-Scott) [reply packages/loader/driver-utils/src/index.ts]: Found `OdspRedirectError` on `packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts`, so using that instead!

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts [hunk 6]
```diff
@@ -227,9 +227,12 @@ export function createOdspNetworkError(
 			break;
 		case 401:
 		case 403:
-			// The server throws 403 status code with innerMostError code as "serviceReadOnly" for cases where the
-			// database on server becomes readonly. The driver retries for such cases with exponential backup logic.
-			if (innerMostErrorCode === OdspServiceReadOnlyErrorCode) {
+			// After a redirect due to a domain change, auth headers are not sent by the browser and can trigger a 403.
+			if (response?.redirected) {
```
[the remark this answers is not in the document]
Comment (Philip-Scott) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:231]: The 403 will be the response on the request against the new domain, and yes `response.url` contains the new URL rather than the old one on this scenario.

On these calls, neither the 308 nor 403 responses contain those redirectLocation error codes. 403 is a `System.UnauthorizedAccessException: Access is denied` error

**Comment by agarwal-navin** [comment]
@vladsud @pragya91 Can you guys please take a look at this PR? @jatgarg is not available right now and you both have the most context here.

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 7]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+}
+
+async function fetchWithLocationRedirectionHandling<T>(
+	getToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+	odspUrlParts: IOdspUrlParts,
+	logger: ITelemetryLoggerExt,
+	options: TokenFetchOptionsEx,
+	fileItem: FileItemLite,
+): Promise<IOdspResponse<Response>> {
+	for (;;) {
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/getFileLink.ts:149]: What is the reason for this loop? My understanding from the changes in the code here is that we only want to perform a retry once after we get the 403 error. If this is true, can we structure the code such that we perform retry within the catch block?

```JS
let { url, requestInit } = await getRequestInformation...
try {
    response = await fetchHelper();
}catch(e){
    // 403 error occured and we need to retry with correct headers here
    if(error.redirectLocation){
        { url, requestInit } = = await getRequestInformation...
         response = await fetchHelper();
    }
}
return response;
```

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 8]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+			logger,
+			options,
+			fileItem,
+		);
+
+		try {
+			return await fetchHelper(url, requestInit);
+		} catch (error: any) {
+			if (
+				error.errorType !== DriverErrorTypes.fileNotFoundOrAccessDeniedError ||
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/getFileLink.ts:162]: Checking for `fileNotFoundOrAccessDeniedError` error type seems unrelated to this change.

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 9]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+				!error.redirectLocation
+			) {
+				throw error;
+			}
+
+			const newSiteDomain = new URL(error.redirectLocation).origin;
+			const oldSiteDomain = new URL(odspUrlParts.siteUrl).origin;
+
+			odspUrlParts.siteUrl = odspUrlParts.siteUrl.replace(oldSiteDomain, newSiteDomain);
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/getFileLink.ts:171]: Changing input parameters of a function is not the best programming practice. (Although I do see it in some parts of the existing code already, but let's avoid adding any new ones.) If possible, always return the new value to the caller, and let the caller assign the value as it may.
(Might be irrelevant if we changed the code here to what I suggested above)

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 10]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
 	return fileLink;
 }
 
+async function getRequestInformation(
```
Comment (pragya91) [inline packages/drivers/odsp-driver/src/getFileLink.ts:96]: Recommend adding code docs to explain the purpose of this function.

**Comment by pragya91** [comment]
Let's add a test for these new changes in `getFileLink.spec.ts`. You can mock throwing of the 403 error with `redirected` property in the response and see if the code works as expected.

**Review by pragya91 — CHANGES_REQUESTED** [review]

**Comment by pragya91** [comment]
@Philip-Scott  As far as I remember the `GetSharingInformation` is not a blocking call for `attach` functionality to go through. However, this api is called when host tries to request the getAbsoluteUrl() on the resolver. Just wanted to highlight that, since the PR talks about the "unable to attach" issue.

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 6 above)
Comment (vladsud) [inline packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:231]: Should we assert that response?.redirected => 401 or 403? I.e. that we are not missing cases?

Comment (vladsud) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:231]: Does 404 / @error.redirectLocation case also have redirect property on response?

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 11]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+	const storageTokenFetcher = toInstrumentedOdspTokenFetcher(
+		logger,
+		odspUrlParts,
+		getToken,
+		true /* throwOnNullToken */,
+	);
+	const storageToken = await storageTokenFetcher(options, "GetFileLinkCore");
+	assert(
+		storageToken !== null,
+		0x2bb /* "Instrumented token fetcher with throwOnNullToken = true should never return null" */,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/getFileLink.ts:112]: Every place that calls storageTokenFetcher() asserts that non-null result is returned. Given that's a wrapper produced by driver, can we at least move this assert into toInstrumentedOdspTokenFetcher() itself?
Or better, TokenFetcher type and force callers to change / assert? That's a bit bigger change, but having null on the type does not help.

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 12]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+	// GetFileByUrl() does not understand that format and thus fails. GetFileById() relies on file item
+	// unique guid (sharepointIds.listItemUniqueId) and it works uniformly across Consumer and Commercial.
+	const { url, headers } = getUrlAndHeadersWithAuth(
+		`${
+			odspUrlParts.siteUrl
+		}/_api/web/GetFileById(@a1)/ListItemAllFields/GetSharingInformation?@a1=guid${encodeURIComponent(
+			`'${fileItem.sharepointIds.listItemUniqueId}'`,
+		)}`,
+		storageToken,
+		true,
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/getFileLink.ts:128]: Please add comments with arg names where it's not very clear what is the arg semantics
```ts
true, // forceAccessTokenViaAuthorizationHeader
```

File: packages/drivers/odsp-driver/src/getFileLink.ts [hunk 13]
```diff
@@ -90,6 +93,86 @@ export async function getFileLink(
...
+			return await fetchHelper(url, requestInit);
+		} catch (error: any) {
+			if (
+				error.errorType !== DriverErrorTypes.fileNotFoundOrAccessDeniedError ||
+				!error.redirectLocation
+			) {
+				throw error;
+			}
+
+			const newSiteDomain = new URL(error.redirectLocation).origin;
```
Comment (vladsud) [inline packages/drivers/odsp-driver/src/getFileLink.ts:168]: Is this right? We can get legit 404 because there is no such file at all (no redirection is involved).
We will get here, likely with undefined error.redirectLocation and will crash, no?

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 7 above)
Comment (Philip-Scott) [reply packages/drivers/odsp-driver/src/getFileLink.ts:149]: Redirects can actually be chained together! One redirect may point to another redirect and so on. 

This was actually a pattern taken from `packages/loader/container-loader/src/location-redirection-utilities/resolveWithLocationRedirection.ts`

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 8 above)
Comment (Philip-Scott) [reply packages/drivers/odsp-driver/src/getFileLink.ts:162]: That is actually the error type we get when throwing a OdspRedirectError that I myself throw on odspErrorUtils.ts. 
image

I'm checking for both that error that I throw, as well as that there is a redirect URL available before doing any redirect logic

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 6 above)
Comment (Philip-Scott) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:231]: I have not seen a 404 redirect error for me to test that this is valid and works. Do you know how one could trigger a 404 redirection error?

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 11 above)
Comment (Philip-Scott) [reply packages/drivers/odsp-driver/src/getFileLink.ts:112]: Mmmm that is actually part of the original code before moving it out to its own function. We can probably follow that up on a separate PR to keep this one small, is that alright?

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 13 above)
Comment (Philip-Scott) [reply packages/drivers/odsp-driver/src/getFileLink.ts:168]: I think this is right, no? On line 163 I am checking that this was the error that I threw, and that redirect Location is set before retrying. If neither of those are true, I'm just re-throwing the error

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 11 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/getFileLink.ts:112]: raised https://github.com/microsoft/FluidFramework/pull/20409 to fix it

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 13 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/getFileLink.ts:168]: Yes, looks good!

File: packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts (hunk 6 above)
Comment (vladsud) [reply packages/utils/odsp-doclib-utils/src/odspErrorUtils.ts:231]: Well, the case below (for 404) has code that deals with redirects. I do not know much about how redirects work (i.e. what error codes can be returned) - @jatgarg & @marcmasmsft are rigth people to consult here.

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 8 above)
Comment (vladsud) [reply packages/drivers/odsp-driver/src/getFileLink.ts:162]: Consider doing the following:

```ts

class OdspRedirectError {
	public static errorType = DriverErrorTypes.fileNotFoundOrAccessDeniedError;
	readonly errorType = OdspRedirectError.type;
...
}

if (error?.errorType !== OdspRedirectError .Type || !error?.redirectLocation)  {
    throw error;
}

```

or even better - way more readable.

class OdspRedirectError {
	public static errorType = DriverErrorTypes.fileNotFoundOrAccessDeniedError;
	readonly errorType = OdspRedirectError.type;

        static public isRedirectError(error: unknown) {
             return error?.errorType  === OdspRedirectError.errorType && !!error?.redirect;
        }
...
}

if (!OdspRedirectError(error)) {
    throw error;
}
...
}

Notice that error does not have to be object (i.e. it can be undefined, null, etc. - we never know)

File: packages/drivers/odsp-driver/src/getFileLink.ts (hunk 7 above)
Comment (pragya91) [reply packages/drivers/odsp-driver/src/getFileLink.ts:149]: Please add comments to explain this logic.

**Comment by christiango** [comment]
Looks like this is now being handled with #21277
